### ISSUE TITLE
Inconsistency for parameters in fixpoints

### ISSUE THREAD

**Issue by yannl35133** [coq/coq#18215]
#### Description of the problem

Arguments are not treated the same in toplevel or anonymous fixpoints :

- a toplevel fixpoint must have a parameter (before the `:=`), and structurality is checked only among those parameters
- an anonymous fixpoint needs not have a parameter, and structurality is checked among the first n parameters*, where n is the number of binders before the `:=`

*the nth parameter of a body is the variable associated to the nth lambda of the body with the binders readded, where the successive lambdas may appear only after a weak-head reduction.

In the end, `check_fix` requires that a fixpoint with structural argument k be one where the kth parameter [as defined above] passes the guard checker.

This means that 
```coq
Definition a := fix f (Pad1 := unit) .. (Padk := unit) :=
  (fun (_ : unit) (pad1 .. padn : False) n =>
    match n with 0 => 0 | S n => f pad1 .. padn n end) tt.
```
will be accepted if k > n (also, if `k=0` or `2 <= k <= n` one  gets "cannot guess decreasing argument of fix", if k = 1 <= n one gets the fully explicit "Recursive definition of f is ill-formed..."), but
```coq
Fixpoint f (Pad1 := unit) .. (Padk := unit) :=
  (fun (_ : unit) (pad1 .. padn : False) n =>
    match n with 0 => 0 | S n => f pad1 .. padn n end) tt.
```
will never be accepted ("A fixpoint needs at least one parameter") and even if some `padi` were transferred to the binder list, `n` would never be tested as a decreasing argument.

#### Trace
I traced the issue to : https://github.com/coq/coq/blob/720cf0248f0cbdf921ad6a22f17345fb18afb294/pretyping/pretyping.ml#L784
The list of checked indexes takes all entries in the binder list, instead of only formal parameters.

#### Coq Version
8.19 master
If I'm right on where the issue is, this was introduced in 8.2

**Comment by yannl35133** [comment]
(NB: the definition is not even needed, `fix f {struct z} := (fun _ n => match n with 0 => 0 | S n => f n end) tt` works

**Comment by herbelin** [comment]
You mean several bugs then?
- `{struct z}` accepted but no is `z` around; `{struct Pad1}` accepted but `Pad1` is a let
- confusion in the presence of let-ins (example `a`)
- inconsistency between `Fixpoint` and `fix`

And maybe more?

**Comment by yannl35133** [comment]
https://github.com/coq/coq/blob/720cf0248f0cbdf921ad6a22f17345fb18afb294/interp/constrintern.ml#L2125-L2130
If the reference in the annotation points to nothing in the context, there is no error and it will point to the index after the binders.

That's (only) two issues

**Comment by herbelin** [comment]
Then, are you doing a fix?

**Comment by yannl35133** [comment]
I can do one later, but I don't know about the fact that `check_fix` is laxer as it does not enforce that bodies be lambdas, something that cannot be reached through manual fixpoints after the fix (even if the current "workaround" is dumb)

**Comment by mrhaandi** [comment]
For reference, in Coq 8.13.2 `Definition a := fix f {struct z} := (fun _ n => match n with 0 => 0 | S n => f n end) tt.`
properly results in the error `No parameter named z.`

**Comment by herbelin** [comment]
The silence on `z` apparently comes from 1e063727ddfaeea9cdf2794e23c5524867e0486f (#15581, from 8.16).

**Comment by yannl35133** [comment]
I guess I should open another issue at some point, but this is a direct answer to the linked commit :
```coq
Generalizable Variable A.
Fixpoint fails `(l : list A) {struct l} : unit := tt.
```
```
Recursive definition of fails is ill-formed.
In environment
fails : forall A : Type, list A -> unit
Recursive definition on "Type"
which should be a recursive inductive type.
Recursive definition is: "fun (A : Type) (_ : list A) => tt".
```
I guess `split_at_annot` is fundamentally flawed since it cannot see these generalised variables, this computation should really happen on at least `Glob_term.extended_glob_local_binder list` (if not `rel_context`)

**Comment by herbelin** [comment]
>  guess split_at_annot is fundamentally flawed since it cannot see these generalised variables, this computation should really happen on at least Glob_term.extended_glob_local_binder list (if not rel_context)

Yes.
